pkg/gofr/service/ratelimiter.go

```diff
@@ -0,0 +1,475 @@
+	}
+
+	r.wg.Add(1)
+
+	req := requestWrapper{
+		execute: execute,
+		respCh:  respCh,
+		ctx:     reqCtx,
+		cancel:  reqCancel,
+	}
+
+	select {
+	case r.requests <- req:
+		r.Log(&Log{
+			Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+			ResponseCode: http.StatusAccepted,
+			HTTPMethod:   "REQUEST",
+			URI:          "rate_limiter",
+		})
+
+		return r.handleResponse(ctx, respCh, reqCancel)
+
+	case <-ctx.Done():
+		atomic.AddInt32(&r.currentQueue, -1)
+		r.wg.Done()
+		reqCancel()
+		r.Log(&ErrorLog{
+			Log: &Log{Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+				ResponseCode: http.StatusGatewayTimeout,
+				HTTPMethod:   "REQUEST",
+				URI:          "rate_limiter",
+			},
+			ErrorMessage: fmt.Sprintf("context canceled while enqueueing: %v", ctx.Err()),
+		})
+
+		return nil, fmt.Errorf("context canceled while enqueueing with TraceID %s: %w", traceID, ctx.Err())
+
+	case <-r.done:
+		atomic.AddInt32(&r.currentQueue, -1)
+		r.wg.Done()
+		reqCancel()
+		r.Log(&ErrorLog{
+			Log: &Log{
+				Timestamp: time.Now(), CorrelationID: traceID, ResponseTime: time.Since(startTime).Microseconds(),
+				ResponseCode: http.StatusServiceUnavailable,
+				HTTPMethod:   "REQUEST",
+				URI:          "rate_limiter",
+			},
+			ErrorMessage: "rate limiter shutting down while enqueueing request",
+		})
+
+		return nil, fmt.Errorf("rate limiter shutting down for request with TraceID %s: %w", traceID, errShuttingDown)
+	}
+}
+
+func (r *RateLimiter) handleResponse(ctx context.Context, respCh chan *requestResponse,
+	reqCancel context.CancelFunc) (*http.Response, error) {
+	span := trace.SpanFromContext(ctx)
+	traceID := span.SpanContext().TraceID().String()
+	startTime := time.Now()
+
+	select {
+	case resp := <-respCh:
+		reqCancel()
+
+		if resp.err != nil {
+			r.Log(&ErrorLog{
+				Log: &Log{
+					Timestamp:     time.Now(),
+					CorrelationID: traceID,
+					ResponseTime:  time.Since(startTime).Microseconds(),
+					ResponseCode:  http.StatusInternalServerError,
+					HTTPMethod:    "REQUEST",
+					URI:           "rate_limiter",
+				},
+				ErrorMessage: fmt.Sprintf("request failed: %v", resp.err),
+			})
+
+			return nil, fmt.Errorf("request failed for TraceID %s: %w", traceID, resp.err)
+		}
+
+		return resp.response, resp.err
+
+	case <-ctx.Done():
+		reqCancel()
+		r.Log(&ErrorLog{
+			Log: &Log{
+				Timestamp:     time.Now(),
+				CorrelationID: traceID,
+				ResponseTime:  time.Since(startTime).Microseconds(),
+				ResponseCode:  http.StatusGatewayTimeout,
+				HTTPMethod:    "REQUEST",
+				URI:           "rate_limiter",
+			},
+			ErrorMessage: fmt.Sprintf("context canceled while waiting: %v", ctx.Err()),
+		})
+
+		return nil, fmt.Errorf("context canceled while waiting for request with TraceID %s: %w", traceID, ctx.Err())
+
+	case <-r.done:
```

Comment:
Here with my idea of not having r.done, ypu could use CancelFunc or context.WithCancelCause to raise/catch the right thing